Try to fix urdfs in memory before SDF import

michalpelka opened the pull request:
Some proposition to solve #476  and #472.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,33 @@
+{
+   namespace Utils
+   {
+       //! Modifies in memory URDF to add missing inertia to links, circle navigate SDF error 19
+       //! @param urdf - the in memory URDF to modify
+       //! @returns a list of links that were modified
+       AZStd::vector<AZStd::string> AddMissingInertiaToLink(AZ::rapidxml::xml_node<>* urdf);
+
+       //! Modifies names of links and joints to be unique, circle navigate SDF error 2
+       //! @param urdf - the in memory URDF to modify
+       //! @returns a list of links that were modified
+       AZStd::vector<AZStd::string> ChangeDuplications(AZ::rapidxml::xml_node<>* urdf);
+
+       std::string ModifyURDFInMemory(const std::string& data);
```
I think we should put this option to add missing `<inertial>` tags and rename duplicate links and joints into two new settings that is apart of the [SdfAssetBuilderSettings](https://github.com/o3de/o3de-extras/blob/development/Gems/ROS2/Code/Source/SdfAssetBuilder/SdfAssetBuilderSettings.cpp#L60-L94).

We can source those settings from the [Settings Registry](https://github.com/o3de/o3de-extras/blob/development/Gems/ROS2/Code/Source/SdfAssetBuilder/SdfAssetBuilderSettings.cpp#L106-L114) and they would be available for the user to select via the File Selection Page in the Robot Importer.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.cpp:
```diff
@@ -47,6 +49,9 @@ namespace ROS2::UrdfParser
-        return Parse(xmlStr, parserConfig);
+        // modify in memory
+
+        std::string xmlStrChanged = (ROS2::Utils::ModifyURDFInMemory(xmlStr));
```
We can make the `Parse*` functions accept an `const SdfAssetBuilderSettings&` instance which can be used to select whether the URDF should be modified in memory before importing.

michalpelka commented:
Added option to enable and disable 
![Screenshot from 2023-09-04 14-39-34](https://github.com/o3de/o3de-extras/assets/3209244/c2075945-1293-4d5c-b90e-e972fa03baea)
Added notification to the User:
![Screenshot from 2023-09-04 14-43-13](https://github.com/o3de/o3de-extras/assets/3209244/f8ccc9fb-61f6-46d6-a0fe-4a097c5ba1c3)

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,135 @@
+#include "FixURDF.h"
+#include <AzCore/XML/rapidxml_print.h>
+#include <AzCore/std/containers/set.h>
+#include <iostream>
```
```suggestion
#include <iostream>

```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,135 @@
+#include <iostream>
+namespace ROS2::Utils
+{
+    AZStd::vector<AZStd::string> AddMissingInertiaToLink(AZ::rapidxml::xml_node<>* urdf)
```
This affects multiple Links. Maybe AddMissingInertiaToLinks (plural)?

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,135 @@
+            {
+                if (linkAndJointsName.contains(name->value()))
+                {
+                    auto newName = AZStd::string(name->value()) + "_joint_dup";
```
I am wondering whether it is priority to keep joint names unchanged (as opposed to keeping link names unchanged) since we actually use joint names for joint manipulation. However, we create entity names and derive namespaces based on links. What do you think?

michalpelka replied:
Both are important. Links names affect frame names, other joint names.
Joint names are easier to change IHMO in frame component.

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+{
+    namespace Utils
+    {
+        //! Modifies in memory URDF to add missing inertia to links, circle navigate SDF error 19
```
```suggestion
        //! Modifies a parsed URDF in memory to add missing inertia to links, which prevents SDF error 19.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+    namespace Utils
+    {
+        //! Modifies in memory URDF to add missing inertia to links, circle navigate SDF error 19
+        //! @param urdf - the in memory URDF to modify
```
```suggestion
        //! @param urdf URDF to modify.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+    {
+        //! Modifies in memory URDF to add missing inertia to links, circle navigate SDF error 19
+        //! @param urdf - the in memory URDF to modify
+        //! @returns a list of links that were modified
```
```suggestion
        //! @returns a list of names of links that were modified.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+        //! @returns a list of links that were modified
+        AZStd::vector<AZStd::string> AddMissingInertiaToLink(AZ::rapidxml::xml_node<>* urdf);
+
+        //! Modifies names of links and joints to be unique, circle navigate SDF error 2
```
```suggestion
        //! Modifies names of links and joints to be unique, which means that a set of both contains no duplicates. This prevents SDF error 2.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+        AZStd::vector<AZStd::string> AddMissingInertiaToLink(AZ::rapidxml::xml_node<>* urdf);
+
+        //! Modifies names of links and joints to be unique, circle navigate SDF error 2
+        //! @param urdf - the in memory URDF to modify
```
```suggestion
        //! @param urdf URDF to modify.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+        AZStd::vector<AZStd::string> ChangeDuplications(AZ::rapidxml::xml_node<>* urdf);
+
+        //! Modifies in memory URDF to add missing elements
+        //! @param urdf - the in memory URDF to modify
```
```suggestion
        //! @param urdf URDF to modify.
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+        //! @param urdf - the in memory URDF to modify
+        //! @returns a modified URDF and a list of XML element that were modified
+        AZStd::pair<std::string, AZStd::vector<AZStd::string>> ModifyURDFInMemory(const std::string& data);
+
```
```suggestion
```

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:
```diff
@@ -147,21 +146,36 @@ namespace ROS2
             {
+                if (urdfParsedWithWarnings)
+                {
+                    report += "# " + tr("The URDF was parsed, but it was modified to be compatible") + "\n";
```
```suggestion
                    report += "# " + tr("The URDF was parsed, though results were modified to be compatible with SDFormat") + "\n";
```

adamdbrw replied:
Not sure if this is right - but please add an explanation (compatible with what?)

adamdbrw commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.cpp:
```diff
@@ -116,13 +118,24 @@ namespace ROS2::UrdfParser
         }
 
         std::string xmlStr((std::istreambuf_iterator<char>(istream)), std::istreambuf_iterator<char>());
+        // modify in memory
```
Nit: move to the inside of curly brackets (like this):
```
{  // modify in memory
```

adamdbrw commented on Gems/ROS2/Code/Source/SdfAssetBuilder/SdfAssetBuilderSettings.cpp:
```diff
@@ -90,7 +92,13 @@ namespace ROS2
+                        AZ::Edit::UIHandlers::Default,
+                        &SdfAssetBuilderSettings::m_fixURDF,
+                        "Fix URDF to be compatible with libsdformat",
+                        "When set, fixes the URDF file before importing it. This is useful for fixing URDF files that have missing inertials."
```
```suggestion
                        "When set, fixes the URDF file before importing it. This is useful for fixing URDF files that have missing inertials or duplicate names within links and joints."
```

adamdbrw commented on Gems/ROS2/Code/Source/SdfAssetBuilder/SdfAssetBuilderSettings.h:
```diff
@@ -36,5 +36,7 @@ namespace ROS2
         bool m_urdfPreserveFixedJoints = true;
         // When true, .dae/.stl mesh files are imported into the project folder to allow the AP to process them
         bool m_importReferencedMeshFiles = true;
+        // When true URDF will be fixed to be compatible with SDF
```
```suggestion
        // When true URDF will be fixed to be compatible with SDFormat.
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.cpp:
```diff
@@ -116,13 +118,24 @@ namespace ROS2::UrdfParser
+        // modify in memory
+        if (Utils::IsFileUrdf(filePath) && settings.m_fixURDF)
+        {
+            const auto& [modifiedXmlStr, modifiedElements] = (ROS2::Utils::ModifyURDFInMemory(xmlStr));
+
+            auto result = Parse(modifiedXmlStr, parserConfig);
+            result.m_modifiedURDFTags = modifiedElements;
+            result.m_modifiedURDFContent = modifiedXmlStr;
+            return result;
```
Instead of using `const auto&` to bind to the each element of the pair as a const reference, a normal `auto` can be used, which can take advantage of return value optimization here. Because the types are now value types in this case, they can also be moved to the result structure to prevent an expensive copy of the parsed XML content.
```suggestion
            auto [modifiedXmlStr, modifiedElements] = (ROS2::Utils::ModifyURDFInMemory(xmlStr));

            auto result = Parse(modifiedXmlStr, parserConfig);
            result.m_modifiedURDFTags = AZStd::move(modifiedElements);
            result.m_modifiedURDFContent = AZStd::move(modifiedXmlStr);
            return result;
```

michalpelka replied:
Good point!

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,136 @@
+    {
+        using namespace AZ::rapidxml;
+        AZStd::vector<AZStd::string> modifiedLinks;
+        AZStd::set<AZStd::string> linkAndJointsName;
```
Using an unordered_set should be quicker if there are a lot of duplicate joint names.
Though for small amount of duplicate joint names, this should be an issue.

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:
```diff
@@ -147,21 +146,36 @@ namespace ROS2
             }
             AZStd::string log;
             bool urdfParsedSuccess{ parsedUrdfOutcome };
+            bool urdfParsedWithWarnings{ parsedUrdfOutcome.m_modifiedURDFContent.size() > 0};
```
Should we codify this into a function on the ParseResult struct in UrdfParser.h?
Perhaps a function such as
```
bool UrdfParsedWithModifiedContent() const;
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.h:
```diff
@@ -10,16 +10,17 @@
 #include <AzCore/IO/Path/Path.h>
 #include <AzCore/std/utility/expected.h>
-#include <sdf/Root.hh>
+#include <SdfAssetBuilder/SdfAssetBuilderSettings.h>
```
Actually this include isn't needed in the UrdfParser.h.

The SdfAssetBuilderSettings can be forward declared in this header and then included in the cpp file.
That will help a little bit with compile times.
```suggestion
```

adamdbrw commented:
![image](https://github.com/o3de/o3de-extras/assets/16702721/1977c7ef-1ef7-4967-b5ed-fda9a3cf5ccc)

Suggestions: 
1. Indicate what the modification was (for example "Modified names:", "Added Inertial").
2. Make it easy to copy / save the modified URDF.

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,136 @@
+        {
+            xml_node<>* inertial = urdf->document()->allocate_node(node_element, "inertial");
+
+            xml_node<>* mass = urdf->document()->allocate_node(node_element, "mass");
+            mass->append_attribute(urdf->document()->allocate_attribute("value", "1."));
```
According to the URDF spec, this should actually default to a 0 mass if not specified: http://wiki.ros.org/urdf/XML/link . "<inertial> (optional: defaults to a zero mass and zero inertia if not specified)". I know the SDF converter requires this to be specified, but does the SDF converter also require this to be 1 and identity for inertia to avoid the error?

michalpelka replied:
I would not set it zero, since it will cause serious problems down the line in PhysX. I think unit mass and identity inertia is reasonable choice.

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,136 @@
+        for (xml_node<>* link = urdf->first_node("link"); link; link = link->next_sibling("link"))
+        {
+            auto* name = link->first_attribute("name");
+            if (name)
+            {
+                linkAndJointsName.insert(name->value());
+            }
```
Does this need to check for duplicate link names too, or is that deliberately ignored here to let the SDF converter throw the error? If so, maybe add a comment here to explain that, something like "// Duplicate link names are invalid in a URDF file, so they won't be handled here. The SDF converter will detect the error case and throw an error."

michalpelka replied:
Done, change the name of the function to `RenameDuplicatedJoints`

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,136 @@
+            {
+                if (linkAndJointsName.contains(name->value()))
+                {
+                    auto newName = AZStd::string(name->value()) + "_joint_dup";
```
This looks like it will only work for the first name collision. If you have 3 joints all named "x", this looks like it would fail. Perhaps this should have an int that's initialized to 0 before the joint loop, and gets added to the name and incremented each time so that we'd have "x_joint_dup_0", "x_joint_dup_1", etc. You could be extra fancy and store a separate counter for each unique name by turning the unordered_set into an unordered_map, but but honestly it's probably enough to just have a global incrementing counter.

michalpelka replied:
I went fancy road here :).

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+{
+    namespace Utils
+    {
+        //! Modifies a parsed URDF in memory to add missing inertia to links, which prevents SDF error 19.
+        //! @param urdf URDF to modify.
+        //! @returns a list of names of links that were modified.
+        AZStd::vector<AZStd::string> AddMissingInertiaToLinks(AZ::rapidxml::xml_node<>* urdf);
+
+        //! Modifies names of links and joints to be unique, which means that a set of both contains no duplicates. This prevents SDF
+        //! error 2.
+        //! @param urdf URDF to modify.
+        //! @returns a list of links that were modified
+        AZStd::vector<AZStd::string> ChangeDuplications(AZ::rapidxml::xml_node<>* urdf);
+
```
Should these definitions get removed from the header, and then just move the comments into the cpp? I can't see why these need to be separately exposed publicly. Also, if you did that, you could move the rapidxml include into the cpp.

michalpelka replied:
It was removed from the header. Thanks for the suggestion.

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.cpp:
```diff
@@ -50,7 +51,7 @@ namespace ROS2::UrdfParser
         return m_root;
     }
-    const sdf::Root& ParseResult::GetRoot() const &
+    const sdf::Root& ParseResult::GetRoot() const&
```
Is the & after the const a typo?

michalpelka replied:
It seems so, but it is not mine. It is a typo existing in codebase. I ran formatting here, what brought to this PR.

michalpelka replied:
It is already exsting in codebase https://github.com/o3de/o3de-extras/blob/fc0e50e7a40e009110fb401f3add5e83f2cdc263/Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.h#L43C6-L43C47
I've here run clang-format.

adamdbrw replied:
Regardless of source, let's remove the &

lumberyard-employee-dm replied:
No the `const&` is not a typo.
It is a referenced qualifier which allows the GetRoot function to return a `const sdf::Root&` when the ParseResult is a const lvalue reference.

This differs when ParseResult is a non-const rvalue reference, such as a call to `UrdfParser::Parse("<xml", ...).GetRoot()` which would return an rvalue `sdf::Root&&` which would be moved automatically.

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/URDF/UrdfParser.h:
```diff
@@ -83,7 +96,9 @@ namespace ROS2
         //!        The relevant ParserConfig functions for URDF importing are
         //!        URDFPreserveFixedJoint() function to prevent merging of robot links bound by fixed joint
         //!        AddURIPath() function to provide a mapping of package:// and model:// references to the local filesystem
+        //! @paragraph settings structure that contains configuration options for the SDFAssetBuilder
```
```suggestion
        //! @param settings structure that contains configuration options for the SDFAssetBuilder
```

michalpelka commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.h:
```diff
@@ -0,0 +1,35 @@
+        //! @returns a list of links that were modified
+        AZStd::vector<AZStd::string> ChangeDuplications(AZ::rapidxml::xml_node<>* urdf);
+
+        //! Modifies in memory URDF to add missing elements
+        //! @param urdf URDF to modify.
+        //! @returns a modified URDF and a list of XML element that were modified
+        AZStd::pair<std::string, AZStd::vector<AZStd::string>> ModifyURDFInMemory(const std::string& data);
```
Moved to CPP, expanded documentation slightly.

mbalfour-amzn commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,145 @@
+                if (linkAndJointsName.contains(name->value()))
+                {
+                    unsigned int& count = linkAndJointsName[name->value()];
+                    auto newName = AZStd::string::format("%s_dup%d", name->value(), count);
```
This should be %u for an unsigned int.

```suggestion
                    auto newName = AZStd::string::format("%s_dup%u", name->value(), count);
```

lumberyard-employee-dm commented on Gems/ROS2/Code/Source/RobotImporter/FixURDF/FixURDF.cpp:
```diff
@@ -0,0 +1,145 @@
+
+        if (links.size())
+        {
+            AZ_Warning("ROS2", false, "Added missing inertia to links: ");
+            for (auto& link : links)
+            {
+                modifiedElements.push_back(link);
+                AZ_Warning("ROS2", false, "  -> %s", link.c_str());
+            }
+        }
+
+        auto renames = RenameDuplicatedJoints(urdf);
+        if (renames.size())
+        {
+            AZ_Warning("ROS2", false, "Renamed duplicated links and joints: ");
+            for (auto& dup : renames)
+            {
+                modifiedElements.push_back(dup);
+                AZ_Warning("ROS2", false, "  -> %s", dup.c_str());
```
I recommend not outputting any messages here, as the caller gets back all of the links that had missing inertia added and the renamed joints as part of the return value of this function.

The caller themselves can then log the messages, if they please.
This helps prevent console spam and allows the caller to choose to output or suppress the warning messages more easily for their use cases.
